ext/project_data.yaml

```diff
@@ -25,7 +25,7 @@ gem_runtime_dependencies:
   locale: '~> 2.1'
   multi_json: '~> 1.10'
   puppet-resource_api: '~>1.5'
-  concurrent-ruby: ["~> 1.0", "< 1.2.0"]
+  concurrent-ruby: "~> 1.0"
```

Comment:
Seems like this would have to be `~> 1.2` as versions before that still created `java.lang.ThreadLocal` objects that became GC roots that would anchor JRuby interpreters and keep them in memory after Puppet Server tried to replace one of its pool workers.

Reply:
I was under the impression the leak was only when running in multithreaded mode. If that's not the case, then do you know if it was fixed upstream (when using the JRuby thread local var implementation on JRuby)?

Reply:
Copying a comment from slack. If we do need to require 1.2.0 or greater, then we'll need to bump the gem in the puppet-runtime and possible puppetserver. Otherwise, puppet will fail to load.

Reply:
From a quick skim of the Puppet source looking at uses of `Puppet::ThreadLocal`, it appears that we are using these objects in a few places without regard to whether multithreaded is enabled or not. Notably, `Puppet::Pops::PuppetStack` which holds the evaluation stack of the compiler.

I think the aim was to make Puppet thread-safe by default so that multithreaded could be switched on, not to make it thread-safe only when multithreaded is on.

Reply:
Thanks for the review @Sharpie . With the steps you provided over Slack, @joshcooper and I did some tests with the new code and older version of concurrent-ruby (1.1.10). It does not appear that this introduces a regression in the GC behavior, so I think we're good to move forward with this.

Reply:
For posterity, what were the the steps?

Reply:
Good call @justinstoller. @Sharpie recommended:

1. Trigger catalog compilation to exercise `ThreadLocalVar`
2. Flush the JRuby pool via the API: `curl -X DELETE -kv --cert "$(puppet config print hostcert)" --key "$(puppet config print hostprivkey)" https://localhost:8140/puppet-admin-api/v1/jruby-pool`
3. Watch `puppetserver.log` to confirm how many JRuby interpreters are created/active, then run `jmap` to verify that there's not a discrepancy between between the two.

Reply:
Thanks.

I don't understand how 1.1.10 doesn't leak, but if you've confirmed the behavior no longer happens, maybe JRuby is now doing something to clean it up.

Reply:
@Sharpie can I get another review on this when you have a chance so we can merge? Thank you!